### PR TITLE
PG-1487: Added support for USX version 3 (chapter end nodes).

### PR DESCRIPTION
**Pull request by tombogle**
Made UsxNode abstract and added sealed subclasses, with tighter checking to ensure correct use.
Improved documentation

https://jira.sil.org/browse/PG-1487

- - -
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1220)

### CODE REVIEW

**Review by andrew-polk — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1220)* status: all files reviewed, 4 unresolved discussions (waiting on @ermshiperete and @tombogle)

___
*[`CHANGELOG.md` line 23 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAq6UtV3dSenlE0Kq1I:-NAq6UtV3dSenlE0Kq1J:br5vqkl) ([raw file](https://github.com/sillsdev/libpalaso/blob/a8b635891a56297cecd51b021f576ffade931b85/CHANGELOG.md#L23)):*
> ```Markdown
> - [SIL.DblBundle] Added protected method GetAttribute to UsxNode.
> - [SIL.DblBundle] Added sealed subclasses of UsxNode: UsxPara and UsxChar.
> - [SIL.DblBundle] Added property IsChapterStart to UsxChapter.
> ```

I think this needs at least a minor version bump.
I'll leave it to you if the subtle breaking change you described below warrants major. (Likely not)
___
*[`CHANGELOG.md` line 27 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAq6SQ5BH-yLFXQs_8_:-NAq6SQ5BH-yLFXQs_8a:bb7phvt) ([raw file](https://github.com/sillsdev/libpalaso/blob/a8b635891a56297cecd51b021f576ffade931b85/CHANGELOG.md#L27)):*
> ```Markdown
> ### Fixed
> - [SIL.DblBundle] Attempting to construct a UsxNode based on an invalid XmlNode now throws an exception in the constructor in most cases rather than later when properties are accessed.
> - [SIL.DblBundle] Accessing UsxChapter.ChapterNumber on a chapter end node returns the chapter umber (from the eid attribute) instead of throwing an exception.
> ```

typo



_Code quote:_
```Markdown
 umber
```
___
*[`SIL.DblBundle/Usx/UsxNode.cs` line 61 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAq5Y031Axh86oC5saB:-NAq5Y031Axh86oC5saC:b-aqsd9o) ([raw file](https://github.com/sillsdev/libpalaso/blob/a8b635891a56297cecd51b021f576ffade931b85/SIL.DblBundle/Usx/UsxNode.cs#L61)):*
> ```Smalltalk
> 
> 	/// <summary>
> 	/// A USX node that specifies a chapter start or end
> ```

Is this right?



_Code quote:_
```Smalltalk
A USX node that specifies a chapter start or end
```
___
*[`SIL.DblBundle.Tests/Usx/UsxNodeTests.cs` line 90 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAq4crh4KQWpAjb472J:-NAq4crh4KQWpAjb472K:baoqnja) ([raw file](https://github.com/sillsdev/libpalaso/blob/a8b635891a56297cecd51b021f576ffade931b85/SIL.DblBundle.Tests/Usx/UsxNodeTests.cs#L90)):*
> ```Smalltalk
> 	{
> 		/// <summary>
> 		/// Tests that the <see cref="UsxPara"/> constructor throws an <see>ArgumentException</see>
> ```

copy/paste error



_Code quote:_
```Smalltalk
see cref="UsxPara"/>
```

File: SIL.DblBundle.Tests/Usx/UsxNodeTests.cs [hunk 1]
```diff
@@ -1,14 +1,52 @@
...
+		public void Constructor_UnderlyingNodeHasNoAttributes_ThrowsArgumentException()
+		{
+			var xmlDoc = new XmlDocument { PreserveWhitespace = true };
+			xmlDoc.LoadXml("<para>This is some text</para>");
+			var usxDoc = new UsxDocument(xmlDoc);
+			Assert.That(() => new UsxPara(usxDoc.GetChaptersAndParas()[0]),
+				Throws.ArgumentException.With.Message.StartsWith(
+					"Invalid USX node. Must have at least one attribute."));
+		}
+	}
```
Comment (ermshiperete) [inline SIL.DblBundle.Tests/Usx/UsxNodeTests.cs:43]: Could add a test that passing `null` throws an `ArgumentException`

File: SIL.DblBundle/Usx/UsxNode.cs [hunk 2]
```diff
@@ -1,67 +1,147 @@
...
+		/// therefore do not have a StyleTag defined. Rather than returning <c>null</c>, we could
+		/// return "c" or a fictitious "c*" or even throw an exception (as we used to do), but
+		/// <c>null</c> seems to best represent the situation.</remarks>
+		public override string StyleTag => IsChapterStart ? base.StyleTag : null;
+
+		/// <summary>
+		/// Gets whether this node represents the start (as opposed to the end) of a chapter.
+		/// </summary>
+		/// <remarks>Starting with USX v. 3, chapter nodes can have an "eid" attribute (and no
+		/// "number" attribute) to indicate that the end of a chapter. A chapter node should
```
Comment (ermshiperete) [inline SIL.DblBundle/Usx/UsxNode.cs:121]:
```suggestion
		/// "number" attribute) to indicate the end of a chapter. A chapter node should
```

File: SIL.DblBundle/Usx/UsxNode.cs [hunk 3]
```diff
@@ -1,67 +1,147 @@
...
+		/// never have both and "sid" and an "eid" attribute, but to accommodate older versions
```
Comment (ermshiperete) [inline SIL.DblBundle/Usx/UsxNode.cs:122]:
```suggestion
		/// never have both an "sid" and an "eid" attribute, but to accommodate older versions
```

("and" is clearly wrong. I changed it to "an" to match what's below and that _sounds_ right, but from my limited English grammar knowledge I'd have thought that it should be "a". Why is it "an"?)

File: SIL.DblBundle.Tests/Usx/UsxNodeTests.cs [hunk 4]
```diff
@@ -22,23 +60,180 @@ public void TestFixtureSetUp()
...
+			var usxNode = new UsxChapter(m_nodes[m_nodes.Count - 1]);
+			Assert.IsNull(usxNode.StyleTag);
+		}
+
+		/// <summary>
+		/// Tests that IsChapterStart returns <c>true</c> when the object is based on a node that
+		/// does not have an sid or eid attribute.
+		/// </summary>
+		[Test]
+		public void IsChapterStart_StartNoSid_ReturnsTrue()
```
Comment (ermshiperete) [inline SIL.DblBundle.Tests/Usx/UsxNodeTests.cs:172]:
```suggestion
		public void IsChapterStart_StartNoSidNoEid_ReturnsTrue()
```

**Review by ermshiperete — CHANGES_REQUESTED** [review]
Andrew already mentioned the need for a +semver:minor/major

**Review by tombogle — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1220)* status: 1 of 6 files reviewed, 10 unresolved discussions (waiting on @andrew-polk, @ermshiperete, and @tombogle)

___
*[`CHANGELOG.md` line 23 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAq6UtV3dSenlE0Kq1I:-NAt8wJv7xsTc_yg0DuE:bc3yblb) ([raw file](https://github.com/sillsdev/libpalaso/blob/a8b635891a56297cecd51b021f576ffade931b85/CHANGELOG.md#L23)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I think this needs at least a minor version bump.
I'll leave it to you if the subtle breaking change you described below warrants major. (Likely not)
</blockquote></details>

Making UsxNode abstract is a "real" breaking change, so it needs to be major.
___
*[`CHANGELOG.md` line 27 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAq6SQ5BH-yLFXQs_8_:-NAt9K3GCzSSYDsmoB9e:b-886gsx) ([raw file](https://github.com/sillsdev/libpalaso/blob/a8b635891a56297cecd51b021f576ffade931b85/CHANGELOG.md#L27)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

typo
</blockquote></details>

Fixed
___
*[`SIL.DblBundle/Usx/UsxNode.cs` line 61 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAq5Y031Axh86oC5saB:-NAtA3Wp7t16RFsC8p23:b32w12) ([raw file](https://github.com/sillsdev/libpalaso/blob/a8b635891a56297cecd51b021f576ffade931b85/SIL.DblBundle/Usx/UsxNode.cs#L61)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Is this right?
</blockquote></details>

No
___
*[`SIL.DblBundle/Usx/UsxNode.cs` line 121 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAjeKkwX2H4VdqhsVbF-r1-121:-NAtACJd5EIQO4hudSz2:b-886gsx) ([raw file](https://github.com/sillsdev/libpalaso/blob/a8b635891a56297cecd51b021f576ffade931b85/SIL.DblBundle/Usx/UsxNode.cs#L121)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

```suggestion
		/// "number" attribute) to indicate the end of a chapter. A chapter node should
```
</blockquote></details>

Fixed
___
*[`SIL.DblBundle/Usx/UsxNode.cs` line 122 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAjeKkwX2H4VdqhsVbF-r1-122:-NAtAbSLEOmsc39y8fsw:b3x4wtp) ([raw file](https://github.com/sillsdev/libpalaso/blob/a8b635891a56297cecd51b021f576ffade931b85/SIL.DblBundle/Usx/UsxNode.cs#L122)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

```suggestion
		/// never have both an "sid" and an "eid" attribute, but to accommodate older versions
```

("and" is clearly wrong. I changed it to "an" to match what's below and that _sounds_ right, but from my limited English grammar knowledge I'd have thought that it should be "a". Why is it "an"?)
</blockquote></details>

"an" is correct (I'll explain why in a Slack message)
___
*[`SIL.DblBundle.Tests/Usx/UsxNodeTests.cs` line 43 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAjeKkwX2H4VdqhsVbD-r1-43:-NAtFC8BAtQZH7KssgZD:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/a8b635891a56297cecd51b021f576ffade931b85/SIL.DblBundle.Tests/Usx/UsxNodeTests.cs#L43)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Could add a test that passing `null` throws an `ArgumentException`
</blockquote></details>

Done.
___
*[`SIL.DblBundle.Tests/Usx/UsxNodeTests.cs` line 90 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAq4crh4KQWpAjb472J:-NAtF_pz5j_wrJ91cEcp:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/a8b635891a56297cecd51b021f576ffade931b85/SIL.DblBundle.Tests/Usx/UsxNodeTests.cs#L90)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

copy/paste error
</blockquote></details>

Done.
___
*[`SIL.DblBundle.Tests/Usx/UsxNodeTests.cs` line 131 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAjeKkwX2H4VdqhsVbD-r1-131:-NAtN8E3A3ijGcO1MKP-:b-vitlyk) ([raw file](https://github.com/sillsdev/libpalaso/blob/a8b635891a56297cecd51b021f576ffade931b85/SIL.DblBundle.Tests/Usx/UsxNodeTests.cs#L131)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

This is fine. However, the tests might be easier to understand if we'd create the nodes inline instead of calling a method in a different class. Ideally we would do it in each test so that it is obvious what attributes are set, but that might be a bit too noisy (or not).

Alternatively, what about changing `UsxDocumentTests.GetChaptersAndParasForMarkOneContaining2Verses` and pass in the extra content string (or the entire content)?
</blockquote></details>

I agree. I think those tests were kind of trying to also test GetChaptersAndParas, but it wasn't very clear. I have removed the optional parameter from that test method and removed all calls to it from outside its fixture. (It is now private)
___
*[`SIL.DblBundle.Tests/Usx/UsxNodeTests.cs` line 172 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAjeKkwX2H4VdqhsVbD-r1-172:-NAtOECr80-dJHwerszi:b-ts45g) ([raw file](https://github.com/sillsdev/libpalaso/blob/a8b635891a56297cecd51b021f576ffade931b85/SIL.DblBundle.Tests/Usx/UsxNodeTests.cs#L172)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

```suggestion
		public void IsChapterStart_StartNoSidNoEid_ReturnsTrue()
```
</blockquote></details>

I merged the two IsChapterStart methods that expect true and renamed accordingly
___
*[`SIL.DblBundle.Tests/Usx/UsxNodeTests.cs` line 193 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAjeKkwX2H4VdqhsVbD-r1-193:-NAtQN-W1dosPPN_7aqr:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/a8b635891a56297cecd51b021f576ffade931b85/SIL.DblBundle.Tests/Usx/UsxNodeTests.cs#L193)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Comment doesn't match the test
</blockquote></details>

Done.
___
*[`SIL.DblBundle.Tests/Usx/UsxNodeTests.cs` line 214 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAjeKkwX2H4VdqhsVbD-r1-214:-NAtQUAN0__mtnUdXxrT:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/a8b635891a56297cecd51b021f576ffade931b85/SIL.DblBundle.Tests/Usx/UsxNodeTests.cs#L214)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Comment doesn't match the test
</blockquote></details>

Done.

File: SIL.DblBundle/Usx/UsxNode.cs [hunk 5]
```diff
@@ -1,67 +1,146 @@
...
 	/// <summary>
-	/// A USX node that specifies the chapter
+	/// A USX node that specifies the start or end of a run of text marked with a character style
 	/// </summary>
-	public class UsxChapter : UsxNode
+	public sealed class UsxChar : UsxNode
 	{
 		/// <summary>
-		/// Creates a chapter XML node 
+		/// Creates a an object representing a USX paragraph based on a USX node.
```
Comment (ermshiperete) [inline SIL.DblBundle/Usx/UsxNode.cs:82]:
```suggestion
		/// Creates an object representing a USX paragraph based on a USX node.
```

File: SIL.DblBundle/Usx/UsxNode.cs [hunk 6]
```diff
@@ -1,67 +1,146 @@
...
+		public XmlNodeList ChildNodes => m_node.ChildNodes;
+	}
+
+	/// <summary>
+	/// A USX node that represents a paragraph (including poetry lines, etc.)
+	/// </summary>
+	public sealed class UsxPara : UsxNode
+	{
+		/// <summary>
+		/// Creates a an object representing a USX paragraph based on a USX node.
```
Comment (ermshiperete) [inline SIL.DblBundle/Usx/UsxNode.cs:66]:
```suggestion
		/// Creates an object representing a USX paragraph based on a USX node.
```

File: SIL.DblBundle/Usx/UsxNode.cs [hunk 7]
```diff
@@ -1,67 +1,146 @@
...
+	}
 
+	/// <summary>
+	/// A USX node that specifies a chapter start or end
+	/// </summary>
+	public sealed class UsxChapter : UsxNode
+	{
 		/// <summary>
-		/// Gets the chapter number
+		/// Creates a an object representing the start or end of a chapter based
```
Comment (ermshiperete) [inline SIL.DblBundle/Usx/UsxNode.cs:98]:
```suggestion
		/// Creates an object representing the start or end of a chapter based
```

**Review by ermshiperete — APPROVED** [review]
lgtm

**Review by tombogle — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1220)* status: 1 of 6 files reviewed, 13 unresolved discussions (waiting on @andrew-polk, @ermshiperete, and @tombogle)

___
*[`SIL.DblBundle/Usx/UsxNode.cs` line 66 at r2](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAjeKkwX2H4VdqhsVbF-r2-66:-NAtZT5x8OET_lyjbsdf:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/64244d5f756d38d1b36fc1ab867c8fb563fdd687/SIL.DblBundle/Usx/UsxNode.cs#L66)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

```suggestion
		/// Creates an object representing a USX paragraph based on a USX node.
```
</blockquote></details>

Done.
___
*[`SIL.DblBundle/Usx/UsxNode.cs` line 82 at r2](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAjeKkwX2H4VdqhsVbF-r2-82:-NAtZU-8D7AbfeD8-Ply:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/64244d5f756d38d1b36fc1ab867c8fb563fdd687/SIL.DblBundle/Usx/UsxNode.cs#L82)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

```suggestion
		/// Creates an object representing a USX paragraph based on a USX node.
```
</blockquote></details>

Done.
___
*[`SIL.DblBundle/Usx/UsxNode.cs` line 98 at r2](https://reviewable.io/reviews/sillsdev/libpalaso/1220#-NAjeKkwX2H4VdqhsVbF-r2-98:-NAtZUfi7od4vTEXVelI:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/64244d5f756d38d1b36fc1ab867c8fb563fdd687/SIL.DblBundle/Usx/UsxNode.cs#L98)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

```suggestion
		/// Creates an object representing the start or end of a chapter based
```
</blockquote></details>

Done.

**Review by andrew-polk — APPROVED** [review]
> Reviewed 5 of 5 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1220)* status: all files reviewed, 10 unresolved discussions (waiting on @ermshiperete)
